hudi-spark-datasource/hudi-spark-common/src/main/java/org/apache/hudi/commit/BaseDatasetBulkInsertCommitActionExecutor.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.hudi.execution.bulkinsert.NonSortPartitionerWithRows;
+import org.apache.hudi.table.BulkInsertPartitioner;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.table.action.HoodieWriteMetadata;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public abstract class BaseDatasetBulkInsertCommitActionExecutor implements Serializable {
+
+  protected final transient HoodieWriteConfig writeConfig;
+  protected final transient SparkRDDWriteClient writeClient;
+  protected final String instantTime;
+  protected HoodieTable table;
+
+  public BaseDatasetBulkInsertCommitActionExecutor(HoodieWriteConfig config,
+                                                   SparkRDDWriteClient writeClient,
+                                                   String instantTime) {
+    this.writeConfig = config;
+    this.writeClient = writeClient;
+    this.instantTime = instantTime;
+  }
+
+  protected void preExecute() {
+    table.validateInsertSchema();
+    writeClient.startCommitWithTime(instantTime, getCommitActionType());
+    writeClient.preWrite(instantTime, getWriteOperationType(), table.getMetaClient());
+  }
+
+  protected abstract HoodieData<WriteStatus> doExecute(Dataset<Row> records, boolean arePartitionRecordsSorted);
+
+  protected void afterExecute(HoodieWriteMetadata<JavaRDD<WriteStatus>> result) {
+    writeClient.postWrite(result, instantTime, table);
+  }
+
+  private HoodieWriteMetadata<JavaRDD<WriteStatus>> buildHoodieWriteMetadata(HoodieData<WriteStatus> writeStatuses) {
+    HoodieWriteMetadata<JavaRDD<WriteStatus>> hoodieWriteMetadata = new HoodieWriteMetadata<>();
+    if (writeStatuses != null) {
+      hoodieWriteMetadata.setWriteStatuses(HoodieJavaRDD.getJavaRDD(writeStatuses));
+      hoodieWriteMetadata.setPartitionToReplaceFileIds(getPartitionToReplacedFileIds(writeStatuses));
+    }
+    return hoodieWriteMetadata;
+  }
+
+  public final HoodieWriteResult execute(Dataset<Row> records, boolean isTablePartitioned) {
+    if (writeConfig.getBoolean(DataSourceWriteOptions.INSERT_DROP_DUPS())) {
+      throw new HoodieException("Dropping duplicates with bulk_insert in row writer path is not supported yet");
+    }
+
+    boolean populateMetaFields = writeConfig.getBoolean(HoodieTableConfig.POPULATE_META_FIELDS);
+
+    BulkInsertPartitioner<Dataset<Row>> bulkInsertPartitionerRows = getPartitioner(populateMetaFields, isTablePartitioned);
+    boolean shouldDropPartitionColumns = writeConfig.getBoolean(DataSourceWriteOptions.DROP_PARTITION_COLUMNS());
+    Dataset<Row> hoodieDF = HoodieDatasetBulkInsertHelper.prepareForBulkInsert(records, writeConfig, bulkInsertPartitionerRows, shouldDropPartitionColumns);
+
+    table = writeClient.initTable(WriteOperationType.INSERT_OVERWRITE, Option.ofNullable(instantTime));
```

Comment:
Will the operation type not depend on the specific implementation of this base class? I don't see `execute` being overridden in subclasses.

Reply:
Fixed